linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala

```diff
@@ -2417,35 +2427,71 @@ private class FunctionEmitter private (
           fb += wa.RefAsNonNull
 
         case _ =>
-          genTree(expr, AnyType)
-
-          markPosition(pos)
+          def targetTpeCastableThroughJSWasmBoundary: Boolean = targetTpe match {
+            case UndefType | StringType | CharType | LongType => false
+            case _                                            => true
+          }
 
-          targetTpe match {
-            case targetTpe: PrimType =>
-              // TODO Opt: We could do something better for things like double.asInstanceOf[int]
-              genUnbox(targetTpe)
+          def default(): Unit = {
+            // TODO Opt: We could do something better for things like double.asInstanceOf[int]
+            genTree(expr, AnyType)
+            markPosition(pos)
+            genCastFromAnyToImpl(targetTpe, targetWasmType)
+          }
 
-            case _ =>
-              targetWasmType match {
-                case watpe.RefType(true, watpe.HeapType.Any) =>
-                  () // nothing to do
-                case watpe.RefType(targetNullable, watpe.HeapType.Extern) =>
-                  fb += wa.ExternConvertAny
-                  if (!targetNullable)
-                    fb += wa.RefAsNonNull
-                case targetWasmType: watpe.RefType =>
-                  fb += wa.RefCast(targetWasmType)
-                case _ =>
-                  throw new AssertionError(s"Unexpected type in AsInstanceOf: $targetTpe")
-              }
+          if (sourceTpe == AnyType && targetTpeCastableThroughJSWasmBoundary) {
+            // Try and push the cast inside nodes that use a custom JS helper
+            expr match {
+              case expr: JSSelect             => genJSSelect(expr, castTo = targetTpe)
+              case expr: JSFunctionApply      => genJSFunctionApply(expr, castTo = targetTpe)
+              case expr: JSMethodApply        => genJSMethodApply(expr, castTo = targetTpe)
+              case expr: SelectJSNativeMember => genSelectJSNativeMember(expr, castTo = targetTpe)
+              case expr: JSUnaryOp            => genJSUnaryOp(expr, castTo = targetTpe)
+              case expr: JSBinaryOp           => genJSBinaryOp(expr, castTo = targetTpe)
+              case expr: JSGlobalRef          => genJSGlobalRef(expr, castTo = targetTpe)
+              case expr: JSTypeOfGlobalRef    => genJSTypeOfGlobalRef(expr, castTo = targetTpe)
+              case expr: JSSuperMethodCall    => genJSSuperMethodCall(expr, castTo = targetTpe)
```

Comment:
This duplication is quite annoying. I've been thinking about better alternatives, but I get stuck at my understanding of `expectedType`. Basically, I do not understand why the following implementation is wrong (with maybe some more cases in `genAdapt`).

```scala
def genCast(expr: Tree, expectedType: Tree, pos: Position): Type = genTree(expr, expectedType)
```

 Is it that we only allow type widening when using `expectedType` and cast allows narrowing? If yes, does that distinction provide value at such a low level?

Reply:
Just thinking out loud: Could it make sense to add `allowCast` to `genTree` (and maybe even `genAdapt`) to avoid the duplication here (and maybe even some between `genAdapt` and `genCastFromAny`?).

Reply:
I don't think so. If we do that, then we must make `allowCast = true` work for any node. Then we would have a problem in `genAdapt`, as we would not have the knowledge that `generatedType <: expectedType`. If we don't have that knowledge, we need to perform subtyping tests to know whether we need to cast. That would be brittle code, which in addition would be dead code in practice; not a great combination :p.